engine/execution/storehouse/register_store.go

```diff
@@ -0,0 +1,266 @@
+}
+
+var _ execution.RegisterStore = (*RegisterStore)(nil)
+
+func NewRegisterStore(
+	diskStore execution.OnDiskRegisterStore,
+	wal execution.ExecutedFinalizedWAL,
+	finalized execution.FinalizedReader,
+	log zerolog.Logger,
+) (*RegisterStore, error) {
+	// replay the executed and finalized blocks from the write ahead logs
+	// to the OnDiskRegisterStore
+	height, err := syncDiskStore(wal, diskStore, log)
+	if err != nil {
+		return nil, fmt.Errorf("cannot sync disk store: %w", err)
+	}
+
+	// fetch the last executed and finalized block ID
+	finalizedID, err := finalized.FinalizedBlockIDAtHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get finalized block ID at height %d: %w", height, err)
+	}
+
+	// init the memStore with the last executed and finalized block ID
+	memStore := NewInMemoryRegisterStore(height, finalizedID)
+
+	return &RegisterStore{
+		memStore:   memStore,
+		diskStore:  diskStore,
+		wal:        wal,
+		finalized:  finalized,
+		finalizing: atomic.NewBool(false),
+		log:        log.With().Str("module", "register-store").Logger(),
+	}, nil
+}
+
+// GetRegister first try to get the register from InMemoryRegisterStore, then OnDiskRegisterStore
+// 1. below pruned height, and is conflicting
+// 2. below pruned height, and is finalized
+// 3. above pruned height, and is not executed
+// 4. above pruned height, and is executed, and register is updated
+// 5. above pruned height, and is executed, but register is not updated since pruned height
+// It returns:
+//   - (value, nil) if the register value is found at the given block
+//   - (nil, nil) if the register is not found
+//   - (nil, storage.ErrHeightNotIndexed) if the height is below the first height that is indexed.
+//   - (nil, storehouse.ErrNotExecuted) if the block is not executed yet
+//   - (nil, storehouse.ErrNotExecuted) if the block is conflicting iwth finalized block
+//   - (nil, err) for any other exceptions
+func (r *RegisterStore) GetRegister(height uint64, blockID flow.Identifier, register flow.RegisterID) (flow.RegisterValue, error) {
+	reg, err := r.memStore.GetRegister(height, blockID, register)
+	// the height might be lower than the lowest height in memStore,
+	// or the register might not be found in memStore.
+	if err == nil {
+		// this register was updated before its block is finalized
+		return reg, nil
+	}
+
+	prunedError, ok := IsPrunedError(err)
+	if !ok {
+		// this means we ran into an exception. finding a register from in-memory store should either
+		// getting the register value or getting a ErrPruned error.
+		return flow.RegisterValue{}, fmt.Errorf("cannot get register from memStore: %w", err)
+	}
+
+	// if in memory store returns PrunedError, and register height is above the pruned height,
+	// then it means the block is connected to the pruned block of in memory store, which is
+	// a finalized block and executed block, so we can get its value from on disk store.
+	if height > prunedError.PrunedHeight {
```

Comment:
would this always be one bigger than pruned height? how could it be bigger than that?